testing/trino-testing-services/src/test/java/io/trino/testng/services/TestReportAfterMethodNotAlwaysRun.java

```diff
@@ -0,0 +1,451 @@
+                        SubClassAddsNotAlwaysRunWithBaseNotAlwaysRun.class.getMethod("afterMethodSub").toString(),
+                },
+        };
+    }
+
+    private static class NoAfterMethods
+    {
+    }
+
+    private static class AllAlwaysRun
+    {
+        @AfterClass(alwaysRun = true)
+        public void afterClass()
+        {
+        }
+
+        @AfterGroups(alwaysRun = true)
+        public void afterGroup()
+        {
+        }
+
+        @AfterMethod(alwaysRun = true)
+        public void afterMethod()
+        {
+        }
+
+        @AfterSuite(alwaysRun = true)
+        public void afterSuite()
+        {
+        }
+
+        @AfterTest(alwaysRun = true)
+        public void afterTest()
+        {
+        }
+    }
+
+    private static class ClassNotAlwaysRun
+    {
+        @AfterClass
+        public void afterClass()
+        {
+        }
+
+        @AfterGroups(alwaysRun = true)
+        public void afterGroup()
+        {
+        }
+
+        @AfterMethod(alwaysRun = true)
+        public void afterMethod()
+        {
+        }
+
+        @AfterSuite(alwaysRun = true)
+        public void afterSuite()
+        {
+        }
+
+        @AfterTest(alwaysRun = true)
+        public void afterTest()
+        {
+        }
+    }
+
+    private static class GroupNotAlwaysRun
+    {
+        @AfterClass(alwaysRun = true)
+        public void afterClass()
+        {
+        }
+
+        @AfterGroups
+        public void afterGroup()
+        {
+        }
+
+        @AfterMethod(alwaysRun = true)
+        public void afterMethod()
+        {
+        }
+
+        @AfterSuite(alwaysRun = true)
+        public void afterSuite()
+        {
+        }
+
+        @AfterTest(alwaysRun = true)
+        public void afterTest()
+        {
+        }
+    }
+
+    private static class MethodNotAlwaysRun
+    {
+        @AfterClass(alwaysRun = true)
+        public void afterClass()
+        {
+        }
+
+        @AfterGroups(alwaysRun = true)
+        public void afterGroup()
+        {
+        }
+
+        @AfterMethod
+        public void afterMethod()
+        {
+        }
+
+        @AfterSuite(alwaysRun = true)
+        public void afterSuite()
+        {
+        }
+
+        @AfterTest(alwaysRun = true)
+        public void afterTest()
+        {
+        }
+    }
+
+    private static class SuiteNotAlwaysRun
+    {
+        @AfterClass(alwaysRun = true)
+        public void afterClass()
+        {
+        }
+
+        @AfterGroups(alwaysRun = true)
+        public void afterGroup()
+        {
+        }
+
+        @AfterMethod(alwaysRun = true)
+        public void afterMethod()
+        {
+        }
+
+        @AfterSuite
+        public void afterSuite()
+        {
+        }
+
+        @AfterTest(alwaysRun = true)
+        public void afterTest()
+        {
+        }
+    }
+
+    private static class TestNotAlwaysRun
+    {
+        @AfterClass(alwaysRun = true)
+        public void afterClass()
```

Comment:
nit: ` public void afterClass() {}` everywhere, the file will be shorter

Reply:
Done (but only methods, I left empty classes alone)